File: Interview Problems/Parking Lot/Java Code/repository/SlotRepository.java
package repository;

import domain.ParkingSlot;
import domain.Vehicle;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

public class SlotRepository {

    private final Map<UUID, ParkingSlot> store = new HashMap<>();

    public void save(ParkingSlot slot) {
        store.put(slot.getId(), slot);
        System.out.println("[REPO] ParkingSlot saved: " + slot.getId());
    }

    public Optional<ParkingSlot> findById(UUID slotId) {
        return Optional.ofNullable(store.get(slotId));
    }

    /**
     * Finds the first available (unoccupied) slot matching the given vehicle type.
     * In production this must be wrapped in a DB transaction with SELECT FOR UPDATE
     * to prevent two concurrent threads allocating the same slot.
     */
    public Optional<ParkingSlot> findAvailableSlot(Vehicle.VehicleType vehicleType) {
        return store.values().stream()
                .filter(s -> s.getSlotType() == vehicleType && !s.isOccupied())
                .findFirst();
    }

    public List<ParkingSlot> findAllOccupied() {
        return store.values().stream()
                .filter(ParkingSlot::isOccupied)
                .collect(Collectors.toList());
    }

    public void update(ParkingSlot slot) {
        if (!store.containsKey(slot.getId())) {
            throw new IllegalArgumentException("Slot not found: " + slot.getId());
        }
        store.put(slot.getId(), slot);
    }
}

File: Interview Problems/Parking Lot/Python Code/repository/slot_repository.py
import uuid
from typing import Dict, List, Optional
from domain.vehicle import VehicleType
from domain.parking_slot import ParkingSlot

class SlotRepository:
    def __init__(self):
        self._store: Dict[uuid.UUID, ParkingSlot] = {}

    def save(self, slot: ParkingSlot):
        self._store[slot.id] = slot
        print(f"[REPO] ParkingSlot saved: {slot.id}")

    def find_by_id(self, slot_id: uuid.UUID) -> Optional[ParkingSlot]:
        return self._store.get(slot_id)

    def find_available_slot(self, vehicle_type: VehicleType) -> Optional[ParkingSlot]:
        for slot in self._store.values():
            if slot.slot_type == vehicle_type and not slot.is_occupied:
                return slot
        return None

    def find_all_occupied(self) -> List[ParkingSlot]:
        return [slot for slot in self._store.values() if slot.is_occupied]

    def update(self, slot: ParkingSlot):
        if slot.id not in self._store:
            raise ValueError(f"Slot not found: {slot.id}")
        self._store[slot.id] = slot
